feat(rule): Adding landmark-is-unique rule

waabid opened the pull request:
The rule checks for whether rules are unique. The structure is:

matches + selectors: identify landmarks.
evaluate: passes all nodes and adds role/label data.
after: fails all landmarks that have duplicate roles and labels.

Linked issue: #866

## Reviewer checks

**Required fields, to be filled out by PR reviewer(s)**
- [x] Follows the commit message policy, appropriate for next version
- [x] Has documentation updated, a DU ticket, or requires no documentation change
- [x] Includes new tests, or was unnecessary
- [x] Code is reviewed for security by: @WilcoFiers

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/dequelabs/axe-core?pullRequest=1394) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/dequelabs/axe-core?pullRequest=1394) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: waabid<br/>:x: markreay<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/dequelabs/axe-core?pullRequest=1394) it.</sub>

dylanb commented on lib/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,19 @@
+	if (foundResultIndex === -1) {
+		uniqueLandmarks.push(currentResult);
+	} else {
+		currentResult.result = false;
```
We should be adding the information about the currentResult to the matching item as a `relatedNode` that way we only return one fail for each duplicated role and name and we can see the others directly in the result.

dylanb replied:
@WilcoFiers I believe that we can build a list of related nodes for each element and the call the result object's `relatedNodes` helper function to turn it into the correct type of array - can you confirm?

waabid replied:
I followed the pattern set by access-keys; let me know what you think 😄

dylanb commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,39 @@
+const excludedParentsForHeaderFooterLandmarks = [
```
Based on testing with VO I think this filtering is unnecessary but need more data

dylanb replied:
for the record: we have decided that because this is a best-practice rule, it will follow the [ARIA 1.1 practices](https://www.w3.org/TR/wai-aria-practices-1.1/#aria_landmark) document regardless of treatment by browser/AT combinations

dylanb commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,39 @@
+	}
+
+	if (tagName === 'section' || tagName === 'form') {
+		const label = axe.commons.aria.label(actualNode);
```
This will be slow on large (deep) pages or with lots of landmarks or when both conditions hold. Need to pass the `virtualNode` into `isLandmarkVirtual` and then into here. Where we do not need the `virtualNode`, we can dereference it there.

waabid replied:
Replaced actual node usage with virtual node usage where there seemed to be virtual functions.

dylanb commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,39 @@
+
+	return (role && landmarkRoles.indexOf(role) >= 0) || role === 'region';
+}
+
```
Replace from here on down with

```
return (
    isLandmarkVirtual(virtualNode) && axe.commons.dom.isVisible(node, true)
);
```

waabid replied:
Done.

dylanb replied:
@waabid I am still showing the old code

waabid replied:
Sorry; I think I misunderstood you. I assume the new changes are what you're looking for. 😄

jkodu commented on lib/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,30 @@
+	const foundResultIndex = findIndex(uniqueLandmarks, findMatch);
+	if (foundResultIndex === -1) {
+		uniqueLandmarks.push(currentResult);
+		currentResult.relatedNodes = [];
```
perhaps setting the relatedNodes to an empty array in the check itself, can help negate this step here?

waabid replied:
I followed the accesskey pattern here; this is because I need info from the node itself (and as such the relatedNodes includes itself when we get to after).

Let me know if this makes sense or I am missing something here. 😄

jkodu commented on lib/checks/landmarks/landmark-is-unique.js:
```diff
@@ -0,0 +1,7 @@
+let role = axe.commons.aria.getRole(node);
```
is it worth adding `if(!role) return false`?

waabid replied:
The reason I didn't add it is because we can be certain by the time we get here, we know for sure that it has a role (since matches would filter out other scenarios).

If you still feel strongly that I should add it, I can.

jkodu commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,45 @@
+	const { actualNode } = virtualNode;
+	const landmarkRoles = axe.commons.aria.getRolesByType('landmark');
+	const role = axe.commons.aria.getRole(actualNode);
+	const tagName = actualNode.nodeName.toLowerCase();
```
as a general practise with in `axe-core`, when using `node.nodeName`, we tend to always use `toUpperCase()`, please replace this with `const nodeName = actualNode.nodeName.toUpperCase()`.

We are looking to enhance our `eslint` rules to catch this at the moment. See https://github.com/dequelabs/axe-core/issues/1052 for more information.

You will have to update your usages as well `nodeName === 'HEADER'` and so on.

dylanb replied:
@wahaj reason for this is XHTML documents

waabid replied:
Done 😄

jkodu commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,45 @@
+		return !!label;
+	}
+
+	return (role && landmarkRoles.indexOf(role) >= 0) || role === 'region';
```
exiting early, when !role, would help with the readability and remove the && here.

waabid replied:
Good call! Added.

dylanb commented on lib/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,30 @@
+		uniqueLandmarks.push(currentResult);
+		currentResult.relatedNodes = [];
+		return true;
+	} else {
```
@jkodu I disagree! But there is a problem with it.

dylanb replied:
Correction! @wahaj the rule seems to be working fine and adding the `relatedNodes` correctly in single iframe pages.

```
"{
 "id": "landmark-unique",
 "impact": "moderate",
 "tags": [
  "cat.semantics",
  "best-practice"
 ],
 "description": "Ensures landmarks are unique",
 "help": "Landmarks must have a unique role or role/label combination(aria-label OR aria-labelledby)",
 "helpUrl": "https://dequeuniversity.com/rules/axe/3.1/landmark-unique?application=axeAPI",
 "nodes": [
  {
   "any": [
    {
     "id": "landmark-is-unique",
     "data": {
      "role": "region",
      "label": "not unique"
     },
     "relatedNodes": [
      {
       "html": "<div role=\"region\" aria-label=\"not unique\">Second occurence</div>",
       "target": [
        "main > div"
       ]
      }
     ],
     "impact": "moderate",
     "message": "The landmarks do not have a unique role or role/label combination (use aria-label OR aria-labelledby to make landmarks distinguishable)."
    }
   ],
   "all": [],
   "none": [],
   "impact": "moderate",
   "html": "<div role=\"region\" aria-label=\"not unique\">First occurence</div>",
   "target": [
    "body > div"
   ],
   "failureSummary": "Fix any of the following:\n  The landmarks do not have a unique role or role/label combination (use aria-label OR aria-labelledby to make landmarks distinguishable)."
  }
 ]
}"
```

dylanb commented on test/integration/rules/landmark-unique/frame.html:
```diff
@@ -0,0 +1,11 @@
+		<script src="/axe.js"></script>
+	</head>
+	<body>
+		<main id="violation-main-2">Second main</main>
```
For this test can you add two landmarks, one with only a role and one with a non-empty name and a role. That way we can be certain that the name is actually being passed up the iframe chain.

dylanb replied:
@waabid I should have been clearer, we also need an assertion on the actual values. I did not see one. If I missed, it, please point it out

waabid replied:
Sorry, I may not be following?

FTR; I added "violation-form-through-iframe-1", which should only fail if the role and name are being passed through iframes (and match the element in the parent element), as an example of when we are indirectly asserting for the values. For direct assertion, I think that might be another form of test (based off of what I've seen in integration tests): is that what you're asking for?

dylanb commented on test/rule-matches/landmark-unique-matches.js:
```diff
@@ -0,0 +1,94 @@
+	});
+
+	(shadowSupport ? it : xit)(
+		'return true for landmarks contained within shadow dom',
```
You should add some tests to ensure that the hierarchical exclusions are working across shadow DOM boundaries

waabid commented on lib/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,30 @@
+		);
+	};
+
+	const findIndex = (array, matchingFunction) => {
```
I actually just used find instead (which is there in polyfills). 😄 

This was more for me to validate that this was the issue anyways.

WilcoFiers commented on lib/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,20 @@
+const uniqueLandmarks = [];
+return results.filter(currentResult => {
```
This solution isn't going to work with scoped runs. This check is assuming that the scope of the run is always the entire page, rather than some element or section on the page. Take this example:

```html
<nav aria-label="foo" id="foo"></nav>
<nav aria-label="foo" id="bar"></nav>
<script> axe.run('#foo') </script>
```

Because `#bar` is out of scope when running axe, it will not be in the `results` array in the after method, and so `#foo` won't be seen as a duplicate. The way to deal with this is to have the `evaluate` method look for duplicates in the page. Have a look at the duplicate-id rule for an example. That also avoids the need of updating relatedNodes in the after method, which while it does work, isn't how this is supposed to work.

General rule of thumb is that when you test the page as a whole, the total number of issues should be the same as when you test separate pieces of the page and concatenate the results.

WilcoFiers replied:
After some extensive discussions, I'm going to withdraw this comment. Dylan and I can't come up with a way to have this rule work across frames, and with limited scope. At least not one that doesn't require substantial changes to the engine.

We think testing landmarks across frames is more important than comparing landmarks outside the scope of the run. So lets go with the proposed solution.

There are other comments to resolve though, so please pick those up, and leave this one for what it is.

waabid replied:
Resolving as such.

WilcoFiers commented on lib/checks/landmarks/landmark-is-unique.json:
```diff
@@ -0,0 +1,16 @@
+	"evaluate": "landmark-is-unique.js",
+	"after": "landmark-is-unique-after.js",
+	"tags": ["best-practice"],
+	"options": {
```
What was the reason for this options object? I don't see this being used anywhere so it looks like this should be removed.

waabid replied:
Removed.

WilcoFiers commented on lib/checks/landmarks/landmark-is-unique.json:
```diff
@@ -0,0 +1,16 @@
+		"impact": "moderate",
+		"messages": {
+			"pass": "Landmarks must have a unique role or role/label combination(aria-label OR aria-labelledby).",
+			"fail": "The landmarks do not have a unique role or role/label combination (use aria-label OR aria-labelledby to make landmarks distinguishable)."
```
As before, I'm not sure why this isn't considering title. That certainly seems to work as a way to set the accessible name of a landmark.

waabid replied:
Updated.

WilcoFiers commented on test/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,84 @@
+
+	function createResultWithSameRelatedNodes(result, data) {
+		return {
+			...createResult(result, data),
```
Babel isn't enabled in our tests, so we need to restrict it to ES3 features. This isn't failing because Appveyor is currently disabled, a fix for that landed recently. There's an issue open to enable Babel on our tests, but for now, use `Object.assign`.

waabid replied:
Done.

WilcoFiers commented on test/checks/landmarks/landmark-is-unique.js:
```diff
@@ -0,0 +1,44 @@
+	'use strict';
+
+	var checkContext = new axe.testUtils.MockCheckContext();
+	let fixture;
```
As above, we can't use ES2015 features in tests. Please use `var` instead of `let` and `const`.

waabid replied:
Done.

WilcoFiers commented on test/rule-matches/landmark-unique-matches.js:
```diff
@@ -0,0 +1,142 @@
+describe('landmark-unique-matches', function() {
+	'use strict';
+	let rule;
```
As above, tests need to be valid ES3.

waabid replied:
Done.

waabid commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,41 @@
+		return isHeaderFooterLandmark(virtualNode);
+	}
+
+	if (nodeName === 'SECTION' || nodeName === 'FORM') {
```
I believe in those cases their roles would be form and section since they are explicit (unlike nav in that example, which is implicit).

If my understanding is incorrect, I can definitely make that change.

waabid commented on lib/rules/landmark-unique-matches.js:
```diff
@@ -0,0 +1,41 @@
+	}
+
+	if (nodeName === 'SECTION' || nodeName === 'FORM') {
+		const label = axe.commons.aria.labelVirtual(virtualNode);
```
My bad: this should actually be using the accessible text API as we have previously discussed.

waabid commented on test/integration/rules/landmark-unique/landmark-unique-fail.html:
```diff
@@ -0,0 +1,29 @@
+<main id="violation-main-1">First main</main>
```
Added.

waabid replied:
Added.

waabid commented on lib/checks/landmarks/landmark-is-unique.js:
```diff
@@ -0,0 +1,7 @@
+let role = axe.commons.aria.getRole(node);
+let label = axe.commons.aria.label(node);
```
Updated.

waabid commented on lib/rules/landmark-unique.json:
```diff
@@ -0,0 +1,13 @@
+	"tags": ["cat.semantics", "best-practice"],
+	"metadata": {
+		"description": "Ensures landmarks are unique",
+		"help": "Landmarks must have a unique role or role/label combination(aria-label OR aria-labelledby)"
```
Done

dylanb commented:
@waabid @WilcoFiers how can I help move this PR along?

waabid commented:
> @waabid @WilcoFiers how can I help move this PR along?

I've addressed all comments; had some tests breaking but fixed those. Now the test build is broken due to a dependency I believe.

WilcoFiers commented:
Closing this as a duplicate of #1394

WilcoFiers commented:
Whoops, I meant the other way around.

WilcoFiers commented on lib/checks/landmarks/landmark-is-unique.json:
```diff
@@ -0,0 +1,14 @@
+		"impact": "moderate",
+		"messages": {
+			"pass": "Landmarks must have a unique role or role/label/title (i.e. accessible name) combination",
+			"fail": "The landmarks do not have a unique role or role/label/title combination (use aria-label, aria-labelledby, or title to make landmarks distinguishable)"
```
We don't need to repeat the requirement here. I 

> "fail": "The landmark must have a unique aria-label, aria-labelledby, or title to make landmarks distinguishable"

WilcoFiers commented on lib/checks/landmarks/landmark-is-unique-after.js:
```diff
@@ -0,0 +1,20 @@
+var uniqueLandmarks = [];
+return results.filter(currentResult => {
```
This file could do with a comment or two explaining the unusual filter logic.

jeankaplansky commented:
Rule help for 3.3 required.
